### PR TITLE
Added Mute Option and Solved  #21

### PR DESCRIPTION
**Pull request by aditya113141**

### CODE REVIEW

**Comment by quozl** [comment]
Thanks, but I'm not quite sure what many of your the changes are for.  See [Files changed](https://github.com/sugarlabs/flappy-birds-activity/pull/23/files) for what I'm looking at.

File: welcomescreen.py [hunk 1]
```diff
@@ -105,12 +108,4 @@ def run(self):
...
-
-        # Just a window exception check condition
-        event1 = pygame.event.get()
-        if event1.type == pygame.QUIT:
-            crashed = True
-
-        if crashed:
-            pygame.quit()
-            sys.exit()
+
```
Comment (aditya113141) [inline welcomescreen.py]: Removed the part where events were dequeued for the second time.

File: scorescreen.py [hunk 2]
```diff
@@ -124,11 +126,4 @@ def run(self, gameDisplay, scores):
...
                 sys.exit()
-
-        # Just a window exception check condition
-        event1 = pygame.event.get()
-        if event1.type == pygame.QUIT:
-            crashed = True
-        if crashed == True:
-            pygame.quit()
-            sys.exit()
+
```
Comment (aditya113141) [inline scorescreen.py]: Removed the part where events were dequeued for the second time.

File: main.py [hunk 3]
```diff
@@ -182,16 +187,7 @@ def make(self):
...
-        # Just a window exception check condition
-        event1 = pygame.event.get()
-        if event1.type == pygame.QUIT:
-            crashed = True
-
-        if crashed:
-            pygame.quit()
-            sys.exit()
-
+
```
Comment (aditya113141) [inline main.py]: Removed the part where events were dequeued for the second time.

File: main.py [hunk 4]
```diff
@@ -161,9 +162,13 @@ def make(self):
             if event.type == pygame.KEYUP and event.key == 273 and self.musicflag:
                 self.keytest = 0
                 self.musicflag = False
+            if event.type == pygame.KEYDOWN and event.key == pygame.K_m:
+                self.sound = not self.sound
+
             if not(self.musicflag) and event.type == pygame.KEYDOWN and event.key == 273:
-                pygame.mixer.music.load("assets/sounds/wing.ogg")
-                pygame.mixer.music.play(0, 0)
+                if self.sound:
```
Comment (aditya113141) [inline main.py]: Added condition to check whether sound is muted or not.

File: elements.py [hunk 5]
```diff
@@ -61,16 +61,18 @@ def display(self, gameDisplay, pillarlist, birds, g):
                                                     self.y + self.height + self.gap + 26 + self.pipedown.get_height() / 2))
         if bird_rect.colliderect(pipe1_rect) or bird_rect.colliderect(pipe2_rect) or \
                 bird_rect.colliderect(pillar1_rect) or bird_rect.colliderect(pillar2_rect):
-            g.hit.play(0)
+            if sound:
```
Comment (aditya113141) [inline elements.py:64]: Added condition to check whether sound should be present or not.

File: elements.py [hunk 6]
```diff
@@ -61,16 +61,18 @@ def display(self, gameDisplay, pillarlist, birds, g):
...
-            b.run(g.gameDisplay, g.scores)
+            b.run(g.gameDisplay, g.scores, sound)
             g.initialize()
             g.welcomeflag = 1
             return
         # scores increment
         if(self.x == 399):
             g.scores += 1
-            g.point.play(0)
+            if sound:
```
Comment (aditya113141) [inline elements.py:74]: Added condition to check whether sound should be present or not.

File: elements.py [hunk 7]
```diff
@@ -131,8 +133,9 @@ def jump(self, land1, land2, land1x, land2x, g):
         platform_rect2 = land2.get_rect(
             center=(land2x + land2.get_width() / 2, 600 + land2.get_height() / 2))
         if bird_rect.colliderect(platform_rect1) or bird_rect.colliderect(platform_rect2):
-            g.hit.play(0)
+            if sound:
+                g.hit.play(0)
```
Comment (aditya113141) [inline elements.py]: Added condition to check whether sound is muted or not.

File: elements.py [hunk 8]
```diff
@@ -61,16 +61,18 @@ def display(self, gameDisplay, pillarlist, birds, g):
...
+            b.run(g.gameDisplay, g.scores, sound)
             g.initialize()
             g.welcomeflag = 1
             return
         # scores increment
         if(self.x == 399):
             g.scores += 1
-            g.point.play(0)
+            if sound:
+                g.point.play(0)
```
Comment (aditya113141) [inline elements.py]: Added condition to check whether sound is muted or not.

File: elements.py [hunk 9]
```diff
@@ -61,16 +61,18 @@ def display(self, gameDisplay, pillarlist, birds, g):
                                                     self.y + self.height + self.gap + 26 + self.pipedown.get_height() / 2))
         if bird_rect.colliderect(pipe1_rect) or bird_rect.colliderect(pipe2_rect) or \
                 bird_rect.colliderect(pillar1_rect) or bird_rect.colliderect(pillar2_rect):
-            g.hit.play(0)
+            if sound:
+                g.hit.play(0)
```
Comment (aditya113141) [inline elements.py]: Added condition to check whether sound is muted or not.

File: scorescreen.py [hunk 10]
```diff
@@ -83,9 +83,11 @@ def run(self, gameDisplay, scores):
             for event in pygame.event.get():
                 # totaltime+=timer.tick()
                 if event.type == pygame.QUIT:
-                    crashed = True
+                    pygame.quit()
+                    sys.exit()
                 if event.type == pygame.KEYDOWN:
-                    swoosh.play(0)
+                    if sound:
+                        swoosh.play(0)
```
Comment (aditya113141) [inline scorescreen.py]: Added condition to check whether sound is muted or not.

File: welcomescreen.py [hunk 11]
```diff
@@ -95,6 +96,8 @@ def run(self):
             self.gameDisplay.blit(head3, (480, 320))
             head3 = font2.render(_("to play the game"), 1, (white))
             self.gameDisplay.blit(head3, (480, 340))
+            mute =  font2.render(_("Press M to Mute/Unmute"),1,(white))
+            self.gameDisplay.blit(mute,(445,550))
```
Comment (aditya113141) [inline welcomescreen.py]: Added instructions for Muting

File: elements.py [hunk 12]
```diff
@@ -34,7 +34,7 @@ def __init__(self):
         self.pipedown_width = self.pipedown.get_width()
         self.pipedown_height = self.pipedown.get_height()
 
-    def display(self, gameDisplay, pillarlist, birds, g):
+    def display(self, gameDisplay, pillarlist, birds, sound, g):
```
Comment (aditya113141) [inline elements.py]: Passed a boolean variable which indicates whether sound is muted or not

File: elements.py [hunk 13]
```diff
@@ -61,16 +61,18 @@ def display(self, gameDisplay, pillarlist, birds, g):
                                                     self.y + self.height + self.gap + 26 + self.pipedown.get_height() / 2))
         if bird_rect.colliderect(pipe1_rect) or bird_rect.colliderect(pipe2_rect) or \
                 bird_rect.colliderect(pillar1_rect) or bird_rect.colliderect(pillar2_rect):
-            g.hit.play(0)
+            if sound:
+                g.hit.play(0)
             b = scorescreen()
-            b.run(g.gameDisplay, g.scores)
+            b.run(g.gameDisplay, g.scores, sound)
```
Comment (aditya113141) [inline elements.py]: Passed a boolean variable which indicates whether sound is muted or not

File: elements.py [hunk 14]
```diff
@@ -108,7 +110,7 @@ def display(self, gameDisplay, flag):
             else:
                 self.y += 1
 
-    def jump(self, land1, land2, land1x, land2x, g):
+    def jump(self, land1, land2, land1x, land2x, sound, g):
```
Comment (aditya113141) [inline elements.py]: Passed a boolean variable which indicates whether sound is muted or not

File: main.py [hunk 15]
```diff
@@ -127,7 +128,7 @@ def make(self):
 
             # Pillar Display
             for i in self.pillarlist:
-                i.display(self.gameDisplay, self.pillarlist, self.birds, self)
+                i.display(self.gameDisplay, self.pillarlist, self.birds, self.sound, self)
```
Comment (aditya113141) [inline main.py]: Passed a boolean variable which indicates whether sound is muted or not

File: main.py [hunk 16]
```diff
@@ -146,7 +147,7 @@ def make(self):
             if(self.land2x <= -140):
                 self.land2x = 837
             if(self.keyinit == 1):
-                self.birds.jump(land1, land2, self.land1x, self.land2x, self)
+                self.birds.jump(land1, land2, self.land1x, self.land2x, self.sound, self)
```
Comment (aditya113141) [inline main.py]: Passed a boolean variable which indicates whether sound is muted or not

File: main.py [hunk 17]
```diff
@@ -161,9 +162,13 @@ def make(self):
             if event.type == pygame.KEYUP and event.key == 273 and self.musicflag:
                 self.keytest = 0
                 self.musicflag = False
+            if event.type == pygame.KEYDOWN and event.key == pygame.K_m:
+                self.sound = not self.sound
```
Comment (aditya113141) [inline main.py]: When key M is pressed, the mute state changes.

File: scorescreen.py [hunk 18]
```diff
@@ -32,7 +32,7 @@
 
 class scorescreen:
 
-    def run(self, gameDisplay, scores):
+    def run(self, gameDisplay, scores,sound):
```
Comment (aditya113141) [inline scorescreen.py]: Passed a boolean variable which indicates whether sound is muted or not

File: activity.py [hunk 19]
```diff
@@ -60,7 +60,7 @@ def build_toolbar(self):
         stop_button = StopButton(self)
         toolbar_box.toolbar.insert(stop_button, -1)
         stop_button.show()
-
+        toolbar_box.show()
```
Comment (quozl) [inline activity.py]: Why is this moved?

File: main.py [hunk 20]
```diff
@@ -113,7 +113,8 @@ def make(self):
             for event in pygame.event.get():
                 # totaltime+=timer.tick()
                 if event.type == pygame.QUIT:
-                    crashed = True
+                    pygame.quit()
+                    sys.exit()
```
Comment (quozl) [inline main.py]: In the sugargame test activity, QUIT is handled by returning.  Why do you handle it here by calling pygame.quit() and sys.exit()?  Doing so bypasses the normal Sugar activity exit path.

File: main.py (hunk 3 above)
Comment (quozl) [reply main.py]: Yet you have not mentioned that by doing this you are fixing https://github.com/sugarlabs/flappy-birds-activity/issues/21

File: scorescreen.py [hunk 21]
```diff
@@ -83,9 +83,11 @@ def run(self, gameDisplay, scores):
             for event in pygame.event.get():
                 # totaltime+=timer.tick()
                 if event.type == pygame.QUIT:
-                    crashed = True
+                    pygame.quit()
+                    sys.exit()
```
Comment (quozl) [inline scorescreen.py]: Again, per sugargame test activity, that's not the right way to exit.

File: welcomescreen.py [hunk 22]
```diff
@@ -70,7 +70,8 @@ def run(self):
             for event in pygame.event.get():
                 # totaltime+=timer.tick()
                 if event.type == pygame.QUIT:
-                    crashed = True
+                    pygame.quit()
+                    sys.exit()
```
Comment (quozl) [inline welcomescreen.py]: Again.

File: scorescreen.py (hunk 2 above)
Comment (quozl) [reply scorescreen.py]: Again.

File: welcomescreen.py (hunk 1 above)
Comment (quozl) [reply welcomescreen.py]: Again.

**Comment by quozl** [comment]
Thanks, but the explanation of your changes should be in the commit message.  Your existing commit message did not mention many of the things that you had since said.  See [Making commits](https://github.com/sugarlabs/sugar-docs/blob/master/src/contributing.md#making-commits) for guidance.

After you explained the changes I was able to see your purpose more clearly, and realised that you had fixed more than one problem in your one commit.  This makes it harder to review.  Try to use different commits for different fixes, as it makes it easier to write commit messages for them.

As of c4769d7 the work remaining is;
* [ ] return instead of calling exit,
* [ ] new code should comply with flake8,
* [ ] rewrite the commits and use communicative commit messages,

Suggested commit messages;
```
Add mute key feature

Add m key to instructions.  Respond to the m key by toggling sound.
```

```
Repair exit path

Problem: (actually don't yet know why you changed this, so looking forward to your words)

Solution: 
```

```
Fix keys are missing

Some key events were missed, and the game would not respond.

Events were dequeued twice in the event loop.

Fixes: URL
```

File: activity.py (hunk 19 above)
Comment (aditya113141) [reply activity.py]: It did not change anything actually. I was playing around with it. This change won't make any difference.

File: main.py (hunk 3 above)
Comment (aditya113141) [reply main.py]: I mentioned it in the pull request title.

Comment (quozl) [reply main.py]: Yes. I know.  But we don't keep the pull request title in git.  We keep the commit messages.  That's why commit messages are to describe and communicate your work.

File: activity.py (hunk 19 above)
Comment (quozl) [reply activity.py]: Okay, thanks.  Are you going to remove this part of the change or would you like someone else to do it?

Comment (aditya113141) [reply activity.py]: Sorry for the delay in response, I had exams. I am working on it currently.

**Comment by aditya113141** [comment]
I will close this and open  new one. I messed up somewhere.

**Comment by quozl** [comment]
https://github.com/sugarlabs/flappy-birds-activity/pull/24
